sigs/jvm/rfcs/20191023-tensorflow-keras.md

```diff
@@ -0,0 +1,342 @@
+    tensorflow operations. For example, in a Dense layer, this method runs the computation
+    *f*(*w_k* x + *w_b*).
+
+    As for `build`, the `call` method overridden within a Layer subclass is wrapped by an `apply`
+    method meant for use external to the layer class:
+    ```java
+    // Override this method in a layer subclass
+    protected abstract Operand<T> call(Ops tf, Operand<T>... inputs)
+
+    // Call this method when using a created Layer object externally.
+    // Performs additional checks on the input length / sizes.
+    public final Operand<T> apply(Ops tf, Operand<T>... inputs)
+    ```
+3. `Layer.computeOutputShape(Shape inputShape)`
+
+    Ensuring tensor dimensions are correct through a sequence of transformations
+    is difficult. The `computeOutputShape` method is used to verify that the transformed
+    output of `apply` has the correct shape.
+
+#### Creating Layers
+
+Standard and user-defined layers might require arguments that are optional, and have defaults.
+To allow for this, the standard layer types each define `Options` and `Builder` classes which allow
+the specification of default and optional arguments. Since using an `Options` class for
+every layer can be cumbersome, static helper functions for creating standard layers
+are defined in the `Layers` class.
+
+Creating a `Dense` layer:
+```java
+// Using Layer Options Builder
+Layer<Float> dense = new Dense<>(128, Dense.Options.builder()
+    .setActivation(Activations.relu)
+    .setKernelInitializer(Initializers.randomNormal)
+    .setBiasInitializer(Initializers.zeros)
+    .build())
+
+// Using static helper
+Layer<Float> dense = Layers.dense(128, Activations.relu, Initializers.randomNormal, Initializers.zeros)
+```
+
+
+## Detailed Design
+
+### Managing the Graph and Ops Accessors
+To use TensorFlow, we must have access to a `Graph` to build and run computations. In Python, this is constructed implicitly. In Java, the `Graph` and `Ops` objects must be created explicitly. Thus, we
+leave this (below) for the user to write, and allow the `Ops` object to be passed
+throughout keras `Layer` construction calls to provide access to core TensorFlow operations.
+
+```java
+try(Graph graph = new Graph) {
+    Ops tf = Ops.create(graph);
+    
+    // Keras code here.
+}
+```
+
+When using the Keras API, models are "compiled" and "fitted" within this block, and the `tf` object is passed through the internal method calls.
+
+### Keras Identifiers
+Standard Keras objects are very often referenced just by name, and don't all have
+to be created explicitly by the user.
+For example, in Python, the strings `"sgd"` or `"adam"` can be used to create different
+kinds of `Optimizer`s, `"relu"` and `"softmax"` stand for specific `Activations`, etc.
+
+In Java, the supported standard types are enumerated in various enum classes: `Activations`, 
+`Intializers`, `Optimizers`, etc., so these types become `Optimizers.sgd`, `Activations.relu`,
```

Comment:
Java standard is to have uppercase labels in enum, e.g. `Activations.RELU`, etc.